0005-ordered-quantification.md

```diff
@@ -0,0 +1,644 @@
+- Feature Name: ordered_quantification
+- Start Date: 2022-02-22 (twosday!)
+- RFC PR: [dafny-lang/rfcs#10](https://github.com/dafny-lang/rfcs/pull/10)
+- Dafny Issue: [dafny-lang/dafny#1753](https://github.com/dafny-lang/dafny/issues/1753)
+
+# Summary
+[summary]: #summary
+
+This RFC proposes adding two closely-related new features to Dafny for working with collections and iteration:
+
+1. `foreach` loops over any enumerable domain, initially only supporting the builtin collection types but leaving the door open for user-defined collections.
+2. Generalized sequence comprehension expressions, similar to the existing `set` and `map` comprehension expressions and more flexible than the current `seq(5, i => i*i)` form.
+
+These features both build on the fundamental new concept of *ordered quantification*, and their syntax and semantics overlap substantially.
+
+# Motivation
+[motivation]: #motivation
+
+Loops are notoriously difficult for Dafny users to verify, and coming up with the correct loop invariants to prove a loop correct 
+is particularly challenging. The recent addition of `for` loops was a good step to abstract away from raw `while` loops,
+but only supports consecutive integer indices. They still fall short of the expressiveness and economy of `for` or `foreach` loops in
+many mainstream programming languages, which generally iterate through the contents of a collection of some kind.
+It is a well-established best practice to avoid manual loop indexes wherever possible, as they force the assumption that datatypes
+support efficient random access, and are more likely to be used incorrectly by accident:
+
+```dafny
+// Before:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  for i := 0 to |s| {
+    var left := s[i];
+    for j := 0 to |s| {
+      var right := s[i]; // Whoops!
+      result := result + (left, right);  
+    }
+  }
+}
+
+// After:
+method AllPairs(s: seq<nat>) returns (result: seq<(nat, nat)>) {
+  result := [];
+  foreach left <- s, right <- s {
```

Comment:
As a quick TL;DR as I revise the proposal, your last version is closest to what I'm thinking now. Those examples would be as follows instead:

`left <- s, right <- s | left < right`
or
`left <- s | left % 2 == 0, right <- s | right % 2 == 0`
or
`left <- s | left !in other, right <- s`

i.e. I'm keeping the hard distinction of `x <- c` not being a boolean operator, but a syntax attached directly to declaring the quantified variable. But each quantified variable can have an optional range attached with `|`. I realized that's still backwards compatible, because existing `x1, x2, ..., xN | <Range>` quantifier domains will just be reinterpreted as having a range attached to the final `xN` variable.

Reply:
Very interesting.